yt/frontends/arepo/data_structures.py

```diff
@@ -65,24 +65,47 @@ def _is_valid(cls, filename, *args, **kwargs):
     def _get_uvals(self):
         handle = h5py.File(self.parameter_filename, mode="r")
         uvals = {}
-        missing = False
-        for unit in ["UnitLength_in_cm", "UnitMass_in_g", "UnitVelocity_in_cm_per_s"]:
+        missing = [False] * 3
+        for i, unit in enumerate(
+            ["UnitLength_in_cm", "UnitMass_in_g", "UnitVelocity_in_cm_per_s"]
+        ):
             if unit in handle["/Header"].attrs:
                 uvals[unit] = handle["/Header"].attrs[unit]
+                if unit == "UnitLength_in_cm":
+                    # We assume this is comoving, because in the absence of comoving
+                    # integration the redshift will be zero.
+                    uvals["cmcm"] = 1.0 / uvals[unit]
             else:
                 mylog.warning("Arepo header is missing %s!", unit)
-                missing = True
+                missing[i] = True
         handle.close()
-        if missing:
+        if all(missing):
             uvals = None
-        else:
-            # We assume this is comoving, because in the absence of comoving
-            # integration the redshift will be zero.
-            uvals["cmcm"] = 1.0 / uvals["UnitLength_in_cm"]
         return uvals
 
     def _set_code_unit_attributes(self):
-        self._unit_base = self._get_uvals()
+        arepo_unit_base = self._get_uvals()
+        # This rather convoluted logic is required to ensure that
+        # units which are present in the Arepo dataset will be used
+        # no matter what but that the user gets warned
+        if arepo_unit_base is not None:
```

Comment:
`arepo_unit_base` is defined as what is in the header (since it comes from `self._get_uvals()`.  But if the header has nothing defined, then this conditional automatically gets skipped.  So it seems weird that there are lines like: "Overwriting X in unit_base with what we found in the dataset."  Isn't the unit_base supposed to override whatever is (or isn't) in the dataset?  Or maybe I'm misunderstanding.

Reply:
I think that `unit_base` should define units when there aren't any. `units_override` works for overriding units--it currently doesn't work for Gadget-based data, but we could make it work in a separate PR. I had avoided doing it before because I wasn't sure how the two things would work with each other, but now I think they could.

Reply:
Now that I think about it, we could probably make this simpler by letting `unit_base` do overrides. I'll think about it more.

Reply:
Yeah, I never understood why there was a separate `unit_override` kwarg from the `unit_base` kwarg.  I would think they both serve the same purpose, which is to set the units for a dataset.  I would assume that if a user is specifying a user_base, then the dataset lacks one, or the that user knows better, so it should clobber the existing one?  But maybe that's not a good assumption.  But that's probably for a different PR.  

The only other suggestion I have for this one is that when one currently runs yt to load up a dataset using the `unit_base`, yt gives warnings saying that "Arepo header is missing UnitLength_in_cm!", even when you're setting them properly with the `unit_base`.  It might be worth either suppressing the "missing" warning or adding an additional yt log statement that yt is using the values prescribed in the `unit_base` to alleviate any freakouts by the user.

Reply:
I'm just going to remove that log message because it's not strictly necessary.